usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1220,22 +1220,13 @@ zil_lwb_write_issue(zilog_t *zilog, lwb_t *lwb)
 	tx = dmu_tx_create(zilog->zl_os);
 
 	/*
-	 * Since we are not going to create any new dirty data and we can even
-	 * help with clearing the existing dirty data, we should not be subject
-	 * to the dirty data based delays.
-	 * We (ab)use TXG_WAITED to bypass the delay mechanism.
-	 * One side effect from using TXG_WAITED is that dmu_tx_assign() can
-	 * fail if the pool is suspended.  Those are dramatic circumstances,
-	 * so we return NULL to signal that the normal ZIL processing is not
-	 * possible and txg_wait_synced() should be used to ensure that the data
-	 * is on disk.
+	 * Since we are not going to create any new dirty data and we
+	 * can even help with clearing the existing dirty data, we
+	 * should not be subject to the dirty data based delays. We
+	 * (ab)use TXG_WAITED to bypass the delay mechanism.
 	 */
-	error = dmu_tx_assign(tx, TXG_WAITED);
-	if (error != 0) {
-		ASSERT3S(error, ==, EIO);
-		dmu_tx_abort(tx);
-		return (NULL);
-	}
+	VERIFY0(dmu_tx_assign(tx, TXG_WAITED));
```

Comment:
Can't `dmu_tx_assign` -> `dmu_tx_try_assign` return `EIO` ?
I think that that still can happen if `spa_suspended` is true and `spa_get_failmode(spa) == ZIO_FAILURE_MODE_CONTINUE`, because `TXG_WAITED` != `TXG_WAIT`.

Perhaps, we also need to change the condition from `txg_how != TXG_WAIT` to `txg_how == TXG_NOWAIT` in the relevant clause in  `dmu_tx_try_assign` like you did in `dmu_tx_assign` ?

Reply:
Thanks @avg-I, I think you're right.